tor_ocr/main.py

```diff
@@ -1,284 +1,146 @@
-            continue
-        except ConnectionError:
-            # try the next API in the list, then release from the loop if we
-            # exhaust our options.
-            continue
-        except RequestException as e:
-            # we have a result object here but it's not right.
-            if result is None:
-                logging.warning(
-                    f'Received null object because of a request exception. '
-                    f'Attempted API: {API} | Error: {e}'
-                )
-            else:
-                logging.error(
-                    f'ERROR {result.status_code} with OCR:\n\nHEADERS:\n '
-                    f'{repr(result.headers)}\n\nBODY:\n{repr(result.text)} '
-                )
-
-    if result is None or not result.ok:
-        raise ConnectionError(
-            'Attempted all three OCR.space APIs -- cannot connect!'
-        )
+        yield s[start: (start + n)]
 
-    return result.json()
 
-
-def escape_reddit_links(body):
-    r"""
-    Escape u/ and r/ links in a message so we don't get confused redditors
-    commenting on transcribot.
-    There is no (known) way to escape u/ or r/ (without a preceding slash),
-    so those will also be changed to \/u/ and \/r/.
-    :param body: the text to escape
-    :return: the escaped text
-    """
-    magic = re.compile('(?<![a-zA-Z0-9])([ur])/|/([ur])/')
-    return magic.sub(r'\/\1\2/', body)
+def get_id_from_url(url: str) -> int:
+    """Extract and return the ID from the end of a Blossom URL."""
+    return int(list(filter(None, urlparse(url).path.split("/")))[-1])
 
 
 # noinspection PyShadowingNames
 def run(config):
     time.sleep(config.ocr_delay)
-    new_post = config.redis.lpop('ocr_ids')
-    if new_post is None:
-        logging.debug('No post found. Sleeping.')
-        # nothing new in the queue. Wait and try again.
-        # Yes, I know this is outside a loop. It will be run inside a loop
-        # by tor_core.
-        return
 
-    # We got something!
-    new_post = new_post.decode('utf-8')
-    logging.info(
-        f'Found a new post, ID {new_post}'
-    )
-    url = config.r.submission(id=clean_id(new_post)).url
 
-    try:
-        result = process_image(url)
-    except OCRError as e:
-        logging.warning(
-            'There was an OCR Error: ' + str(e)
-        )
+    new_posts = config.blossom.get_ocr_transcriptions().data
+    if len(new_posts) == 0:
+        logging.debug("No new posts found; sleeping.")
         return
 
-    logging.debug(f'result: {result}')
+    logging.info(f"Retrieved {len(new_posts)} unprocessed posts")
 
-    if not result:
-        logging.info('Result was none! Skipping!')
-        # we don't want orphan entries
-        config.redis.delete(new_post)
-        return
-
-    tor_post_id = config.redis.get(new_post).decode('utf-8')
+    for post in new_posts:
+        # there should be two transcriptions: one from the volunteer
```

Comment:
Is this necessarily the case? If a post has just been submitted, there is not necessarily a transcription by the volunteer right?

Reply:
Excellent point, that is true